ragna/assistants/_cohere.py

```diff
@@ -0,0 +1,81 @@
+from typing import AsyncIterator, cast
+
+from ragna.core import RagnaException, Source
+
+from ._api import ApiAssistant
+
+
+class CohereApiAssistant(ApiAssistant):
+    _API_KEY_ENV_VAR = "COHERE_API_KEY"
+    _MODEL: str
+    _CONTEXT_SIZE: int = 4_000
+
+    @classmethod
+    def display_name(cls) -> str:
+        return f"Cohere/{cls._MODEL}"
+
+    @property
+    def max_input_size(self) -> int:
+        return self._CONTEXT_SIZE
+
+    def _make_system_content(self) -> str:
+        instruction = (
+            "You are a helpful assistant that answers user questions given the included context. "
+            "If you don't know the answer, just say so. Don't try to make up an answer. "
+            "Only use the included documents below to generate the answer."
+        )
+        return instruction
+
+    def _make_source_documents(self, sources: list[Source]) -> list[dict[str, str]]:
+        document_sources = [
+            {"title": source.id, "snippet": source.content} for source in sources
+        ]
+        return document_sources
+
+    async def _call_api(
+        self, prompt: str, sources: list[Source], *, max_new_tokens: int
+    ) -> AsyncIterator[str]:
+        response = await self._client.post(
+            "https://api.cohere.ai/v1/chat",
```

Comment:
Cohere API is able to stream the response. This results in a better UX and we should always do that if the API supports it. From the [wording](https://docs.cohere.com/reference/chat), I guess they are using SSE to stream

> the response will be a JSON stream of events

If so, you can have a look at the OpenAI or Anthropic assistant on how to do that. If they for some reason mean that they actually respond with a JSON stream, have a look at the Google assistant on how to handle this.

Reply:
It's a JSON stream. I'm looking now how to do it